lib/iris/common/metadata.py

```diff
@@ -0,0 +1,402 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+
+from abc import ABCMeta
+from collections import namedtuple
+from collections.abc import Iterable, Mapping
+from functools import wraps
+import re
+
+
+__all__ = [
+    "AncillaryVariableMetadata",
+    "BaseMetadata",
+    "CellMeasureMetadata",
+    "CoordMetadata",
+    "CubeMetadata",
+    "MetadataFactory",
+]
+
+
+# https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#object_name
+_TOKEN_PARSE = re.compile(r"""^[a-zA-Z0-9][\w\.\+\-@]*$""")
+
+
+class _BaseMeta(ABCMeta):
+    """
+    Meta-class to support the convenience of creating a namedtuple from
+    names/members of the metadata class hierarchy.
+
+    """
```

Comment:
Reference:
- https://docs.python.org/3/reference/datamodel.html?highlight=namespace#object.__new__
- https://docs.python.org/3/reference/datamodel.html?highlight=namespace#metaclasses
- https://docs.python.org/3/reference/datamodel.html?highlight=namespace#creating-the-class-object